### PR TITLE
support type annot in constants, casts

### LINKED ISSUE #54332: NLL should respect lifetimes in `as` expressions
Spawned off from https://github.com/rust-lang/rust/issues/47184. The following test case compiles, but it should not:

```rust
#![allow(warnings)]
#![feature(nll)]

fn main() {
    let x = 22_u32;
    let y: &u32 = (&x) as &'static u32;
}
```

The problem is that the "cast" expression is not preserving the full type that the user gave (`&'static u32`). 

This is very similar to https://github.com/rust-lang/rust/issues/54331

### LINKED ISSUE #54571: nll: respect user type annotations with constants in expressions
Breaking out from https://github.com/rust-lang/rust/issues/47184 as a subtask:

We need to respect user annotations when referencing constants in expressions. First job: come up with some test cases.

### LINKED ISSUE #55183: Internal compiler error, broken MIR, in latest Rust nightly
The latest Rust nightly (`rustc 1.31.0-nightly (1dceaddfb 2018-10-17)`) has introduced an compiler error, that causes the compiler to crash when compiling `nalgebra`. I'm not entirely sure posting this here is correct. The problem was introduced in a minor Rust nightly compiler update, and if the code is bad, I don't think the Rust compiler should crash the way it does now.

Here is a snippet of errors that are coming up ([complete log](https://gist.github.com/timvisee/3ad2ede950bc9deb7b351a2fd9cc8645)):
```
error: internal compiler error: broken MIR in DefId(0/0:8133 ~ nalgebra[ff9d]::base[0]::swizzle[0]::{{impl}}[0]::xx[0]) (const base::construction::<impl base::matrix::Matrix<N, base::dimension::U2, base::dimension::U1, <base::default_allocator::DefaultAllocator as base::allocator::Allocator<N, base::dimension::U2>>::Buffer>>::new): bad constant user type FnDef(DefId(0/0:8339 ~ nalgebra[ff9d]::base[0]::construction[0]::{{impl}}[47]::new[0]), Canonical { variables: [CanonicalVarInfo { kind: Ty(General) }, CanonicalVarInfo { kind: Ty(General) }, CanonicalVarInfo { kind: Ty(General) }], value: UserSubsts { substs: [_], user_self_ty: Some(UserSelfTy { impl_def_id: DefId(0/0:8338 ~ nalgebra[ff9d]::base[0]::construction[0]::{{impl}}[47]), self_ty: base::matrix::Matrix<_, base::dimension::U2, base::dimension::U1, base::matrix_array::MatrixArray<_, base::dimension::U2, base::dimension::U1>> }) } }) vs fn(N, N) -> base::matrix::Matrix<N, base::dimension::U2, base::dimension::U1, <base::default_allocator::DefaultAllocator as base::allocator::Allocator<N, base::dimension::U2>>::Buffer> {base::construction::<impl base::matrix::Matrix<N, base::dimension::U2, base::dimension::U1, <base::default_allocator::DefaultAllocator as base::allocator::Allocator<N, base::dimension::U2>>::Buffer>><N>::new}: NoSolution
  --> src/base/swizzle.rs:14:30
   |
14 |                       $Result::new($(self[$i]),*)
   |                                ^^^
...
21 | / impl_swizzle!(
22 | |     where U0: xx() -> Vector2[0, 0];
23 | |     where U1: xy() -> Vector2[0, 1];
24 | |     where U2: xz() -> Vector2[0, 2];
...  |
66 | |     where U2: zzz() -> Vector3[2, 2, 2]
67 | | );
   | |__- in this macro invocation

error: internal compiler error: broken MIR in DefId(0/0:8135 ~ nalgebra[ff9d]::base[0]::swizzle[0]::{{impl}}[1]::xy[0]) (const base::construction::<impl base::matrix::Matrix<N, base::dimension::U2, base::dimension::U1, <base::default_allocator::DefaultAllocator as base::allocator::Allocator<N, base::dimension::U2>>::Buffer>>::new): bad constant user type FnDef(DefId(0/0:8339 ~ nalgebra[ff9d]::base[0]::construction[0]::{{impl}}[47]::new[0]), Canonical { variables: [CanonicalVarInfo { kind: Ty(General) }, CanonicalVarInfo { kind: Ty(General) }, CanonicalVarInfo { kind: Ty(General) }], value: UserSubsts { substs: [_], user_self_ty: Some(UserSelfTy { impl_def_id: DefId(0/0:8338 ~ nalgebra[ff9d]::base[0]::construction[0]::{{impl}}[47]), self_ty: base::matrix::Matrix<_, base::dimension::U2, base::dimension::U1, base::matrix_array::MatrixArray<_, base::dimension::U2, base::dimension::U1>> }) } }) vs fn(N, N) -> base::matrix::Matrix<N, base::dimension::U2, base::dimension::U1, <base::default_allocator::DefaultAllocator as base::allocator::Allocator<N, base::dimension::U2>>::Buffer> {base::construction::<impl base::matrix::Matrix<N, base::dimension::U2, base::dimension::U1, <base::default_allocator::DefaultAllocator as base::allocator::Allocator<N, base::dimension::U2>>::Buffer>><N>::new}: NoSolution
  --> src/base/swizzle.rs:14:30
   |
14 |                       $Result::new($(self[$i]),*)
   |                                ^^^
...
21 | / impl_swizzle!(
22 | |     where U0: xx() -> Vector2[0, 0];
23 | |     where U1: xy() -> Vector2[0, 1];
24 | |     where U2: xz() -> Vector2[0, 2];
...  |
66 | |     where U2: zzz() -> Vector3[2, 2, 2]
67 | | );
   | |__- in this macro invocation

# --- snip ---
```

This error is coming up when `check`ing, `build`ing and when `build`ing for `--release`.

I'm running this on `Ubuntu 18.04`, here's some more information:
```bash
$ rustc -Vv
rustc 1.31.0-nightly (1dceaddfb 2018-10-17)
binary: rustc
commit-hash: 1dceaddfbe163e2d916c904b98923342730ba970
commit-date: 2018-10-17
host: x86_64-unknown-linux-gnu
release: 1.31.0-nightly
LLVM version: 8.0

$ cargo -Vv
cargo 1.31.0-nightly (5dbac9888 2018-10-08)
release: 1.31.0
commit-hash: 5dbac98885199bbd7c0f189d7405b5523434d1e3
commit-date: 2018-10-08

$ uname -srmo
Linux 4.15.0-33-generic x86_64 GNU/Linux
```

To reproduce this, ensure you've the same Rust nightly version installed as mentioned above. Then clone the [`nalgebra`](https://github.com/rustsim/nalgebra) repository and run `cargo check` (faster than compiling). The error should come up when checking `nalgebra` starts.

### PR DESCRIPTION
**Pull request by nikomatsakis**
Fixes #54571 
Fixes #54332
Fixes #55183

r? @pnkfelix

### CODE REVIEW

**Comment by nikomatsakis** [comment]
OK, works now.

**Comment by nikomatsakis** [comment]
Well, famous last words. =)

**Comment by nikomatsakis** [comment]
@pnkfelix I threw some more commits in here, fixing up cast expressions (#54332)

**Comment by nikomatsakis** [comment]
OK so this had some fallout on diagnostics -- I think because we are proving some things multiple times? We can probably find a way to suppress the duplicates.

The constant errors I am less sure how to fix (cc @oli-obk @RalfJung). The underlying problem is that with a cast like `X as usize` we will now sometimes generate temporaries like:

```
T1 = X
AscribeUserType(T1, usize)
```

Perhaps the best fix would be to do this ascription less often, actually. For example, if there are no regions in type we are casting to, we don't have to preserve the user-supplied type (NLL doesn't care).

**Comment by nikomatsakis** [comment]
OK, I pushed a workaround and fixed some other bugs -- there is one FIXME I am still pondering, about what we prove WF for, but this PR is strictly better than the status quo and (I think) probably fine as is. The remaining potential bug (that I know of) probably exists without NLL and is really about normalization and hence more the domain of https://github.com/rust-lang/rust/issues/54940.

**Comment by RalfJung** [comment]
> The constant errors I am less sure how to fix

Which errors are you referring to?

> The underlying problem is that with a cast like `X as usize` we will now sometimes generate temporaries

What is the type of `X` in this example?

**Comment by oli-obk** [comment]
The type ascription statement is completely irrelevant for const propagation. I propose that you simply skip over it in https://github.com/rust-lang/rust/blob/f7eb7fbbf6bca0eaa146c4a48cd39c133a257706/src/librustc_mir/transform/const_prop.rs#L567 instead of falling through to https://github.com/rust-lang/rust/blob/f7eb7fbbf6bca0eaa146c4a48cd39c133a257706/src/librustc_mir/transform/const_prop.rs#L584 which will then see the constant.

**Comment by nikomatsakis** [comment]
@oli-obk so you are proposing that I replace that `if let` with something like

```rust
match statement.kind {
    Assign => /* as today */,
    Ascription => { }
    _ => self.super_statement(...),
}
```

?

**Comment by oli-obk** [comment]
I think we are depending on `Assign` also executing `super_statement`, so you'll need to copy that also into the `Assign` arm, but otherwise, yes

**Comment by nikomatsakis** [comment]
@oli-obk I'll take a look -- I need to make a test case that still fails, and the rebase seems to have exposed some other diagnostic regression (59ca62f385d3e3527faefa099ddae53f462af2ac). I don't quite see how that particular visitor you pointed me at is related yet but I'm sure it'll be clear if I poke at the test case.

**Comment by nikomatsakis** [comment]
Oh, I think the diagnostic regression is actually my fault for mis-using `--keep-stage 1` =) Heh.

**Comment by nikomatsakis** [comment]
I should probably flatten out this history too it's getting messy

**Comment by oli-obk** [comment]
>  I need to make a test case that still fails,

that might be hard, since the const propagator only propagates primitives, which don't have regions. Since (I think?) you stopped producing `Ascription` statements for such types, the changes I proposed are very likely not necessary anymore.

**Comment by nikomatsakis** [comment]
@oli-obk what about a `&'static u32` value or something?

**Comment by nikomatsakis** [comment]
OK, squashing duplicates fixes one of the error message regressions. Not sure about the other one yet.

**Comment by oli-obk** [comment]
I don't think we propagate references (yet). Idk, doesn't seem too important, we can fix the duplication if it ever comes up in the future

**Comment by nikomatsakis** [comment]
I figured out why the other error winds up being reported. The answer is that the predicates we instantiate from the environment wind up containing named regions and things, and not only the "inference variables" that we get otherwise. This makes them take a slightly different (though equivalent) path that winds up with something that my simple heuristic was not considering to be a "duplicate".

**Comment by nikomatsakis** [comment]
I could do a looser "duplicate" heuristic -- e.g., always consider a report that `T: 'a` and `T::Item: 'a` to be duplicates -- but it seems potentially inaccurate. We could also convert the named regions we extract in the predicates into their equivalent region-vids, which would solve the problem.

**Comment by nikomatsakis** [comment]
Or we could just live with it =)

**Comment by nikomatsakis** [comment]
OK, I think I resolved it.

**Comment by nikomatsakis** [comment]
Added a fix for #55183

**Comment by nikomatsakis** [comment]
Added a fix for #55183

**Review by pnkfelix — COMMENTED** [review]
🎉 for  "skip user-type annotations if they don't have regions" a0b4bef4f56ea8ea27c430d5857ddf53dc92b541

File: src/librustc_typeck/check/mod.rs [hunk 1]
```diff
@@ -2361,9 +2361,13 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         let ty = self.to_ty(ast_ty);
 
         // If the type given by the user has free regions, save it for
-        // later, since NLL would like to enforce those. Other sorts
-        // of things are already sufficiently enforced. =)
-        if ty.has_free_regions() {
+        // later, since NLL would like to enforce those. Also pass in
+        // types that involve projections, since those can resolve to
+        // `'static` bounds (modulo #54940, which hopefully will be
+        // fixed by the time you see this comment, dear reader,
```
Comment (pnkfelix) [inline src/librustc_typeck/check/mod.rs]: 😆

**Comment by nikomatsakis** [comment]
cc @rust-lang/infra — looks like a network error or something — am I supposed to catalog this in some way?

**Comment by kennytm** [comment]
@nikomatsakis Yes and we've recorded this down. There was [an outage in S3 in the US-WEST-1 region around 2018-10-19 17~18 UTC](https://status.aws.amazon.com/) causing of this errpr. Error in deployment is basically always spurious and unfortunately something out of our control.

**Pull request merged**